Add internationalization

tulir opened the pull request:
Fixes #99 

This pull request adds internationalization (I18n) support to everything using aurelia-i18n (I18next). This also adds some of the Finnish translations.

Demo available at https://dev.openapi.design/feature/i18n/

tulir commented:
Note to self: We can probably easily implement internationalization of field labels by simply using the full path to the fields as the I18n key. That way it requires no schema changes and we can use the current labels as fallback if necessary.

tulir commented:
Internationalization is now implemented for most fields, but we need to add at least an English localization.

tulir commented:
There's a continuation of this branch in the next pull request, so don't rebase this to avoid having to manually fix things.

brylie commented on src/app.html:
```diff
@@ -42,16 +42,16 @@
         </button>
         <div class="content">
-          <a href="javascript:void(0)" click.delegate="delete()">Delete</a>
+          <a href="javascript:void(0)" click.delegate="delete()" t="nav.actions.delete">Delete data</a>
```
While we are here, lets nest the child content of this anchor element.

brylie commented on src/app.html:
```diff
@@ -42,16 +42,16 @@
         <div class="content">
-          <a href="javascript:void(0)" click.delegate="download('json')">JSON</a>
-          <a href="javascript:void(0)" click.delegate="download('yaml')">YAML</a>
+          <a href="javascript:void(0)" click.delegate="download('json')" t="nav.actions.download.json">JSON</a>
```
While we are here, lets nest the child content of these anchor elements.

brylie commented on src/app.html:
```diff
@@ -126,15 +126,15 @@
     <div class="error-data">
-      <h1>Uh oh!</h1>
-      <div class="text">
+      <h1 t="error.title">Uh oh!</h1>
```
While we are here, lets nest the child content of this anchor element.

brylie commented on src/app.html:
```diff
@@ -126,15 +126,15 @@
       </div>
 
-      <h3>Exception data</h3>
+      <h3 t="error.data-title">Exception data</h3>
```
While we are here, lets nest the child content of this anchor element.

brylie commented on src/app.html:
```diff
@@ -146,7 +146,7 @@
   <div class="brand-footer">
     <div class="top-row">
-      <span class="powered-by">Powered by</span>
+      <span class="powered-by" t="footer-powered-by">Powered by</span>
```
While we are here, lets nest the child content of this element.

brylie commented on src/app.js:
```diff
@@ -1,15 +1,22 @@
 
     try {
-      this.forms = parseJSON('swagger', JSON.parse(JSON.stringify(schema)));
+      this.forms = parseJSON('form', JSON.parse(JSON.stringify(schema)));
```
Lets use a named variable here, since there are so many levels of nesting.

E.g.

```js
const parsedSchema = JSON.parse(JSON.stringify(schema));

this.forms = parseJSON('form', parsedSchema);
```

brylie commented on src/app.js:
```diff
@@ -1,15 +1,22 @@
-      this.forms = parseJSON('swagger', JSON.parse(JSON.stringify(schema)));
+      this.forms = parseJSON('form', JSON.parse(JSON.stringify(schema)));
       if (window.localStorage['openapi-v2-design']) {
         this.forms.setValue(JSON.parse(window.localStorage['openapi-v2-design']));
```
Lets use a named variable here, since there are so many levels of nesting. E.g.

```js
// Get previous OpenAPI v2 design from local storage
const previousDesign = JSON.parse(window.localStorage['openapi-v2-design']);

// populate form from previous design
this.forms.setValue(previousDesign)
```

brylie commented on src/lang/en.json:
```diff
@@ -0,0 +1,391 @@
+  },
+  "error": {
+    "title": "Uh oh!",
+    "body": "It seems like something is broken.\nSometimes these issues are caused by an invalid OpenAPI document in the cache. If this is the case, <a href=\"javascript:void(0)\" click.delegate=\"delete()\">deleting the data</a> should fix it.",
```
I am not sure that we want to embed so much markup within our localizations. See if you can break this localization into multiple parts or add parameters. Markup should, ideally, live in the template.

tulir replied:
Is there any way to do interpolation with Aurelia in-template I18n?

brylie commented on src/lang/en.json:
```diff
@@ -0,0 +1,391 @@
+      "label": "Paths",
+      "newItemText": "New Path",
+      "item": {
+        "label": "Path #$index: ${#:key}",
```
In this case, we might consider moving these parameters into the template. I am not sure how the grammar of different languages would change their placement in a phrase, but lets consider keeping them in markup for now.

cc: @bajiat

tulir replied:
We could add a section to the schema to define and name the variables a certain localization usually needs to reduce variable inserting noise in the locale files.

tulir replied:
Making direct links (e.g. `{{path}}` -> `/the/path`) would break instant updating, so maybe `{{path}}` -> `${#:key}` would be better.

brylie commented on src/lang/en.json:
```diff
@@ -0,0 +1,391 @@
+            "label": "Tags",
+            "newItemText": "New Tag",
+            "item": {
+              "label": "Tag #$index"
```
Perhaps this placeholder can remain here. Lets just determine a consistent way to handle these types of placeholders. I.e. what would it look like if we had the `#$index` placeholder in the template, or is this an i18next parameter?

brylie replied:
My intuition here is telling me that we might only use i18next [interpolation](https://www.i18next.com/interpolation.html) and [pluralization](https://www.i18next.com/plurals.html), leaving most/all other parameters and markup out of the localization strings.

brylie commented on src/lang/en.json:
```diff
@@ -0,0 +1,391 @@
+              "label": "Scopes",
+              "newItemText": "New Scope",
+              "item": {
+                "label": "Scope #$index",
```
My intuition here is telling me that we might only use i18next [interpolation](https://www.i18next.com/interpolation.html) and [pluralization](https://www.i18next.com/plurals.html), leaving most/all other parameters and markup out of the localization strings.

brylie commented on src/lang/en.json:
```diff
@@ -0,0 +1,391 @@
+        "label": "Requirements",
+        "newItemText": "New Requirements Object",
+        "item": {
+          "label": "Requirements Object #$index",
```
My intuition here is telling me that we might only use i18next [interpolation](https://www.i18next.com/interpolation.html) and [pluralization](https://www.i18next.com/plurals.html), leaving most/all other parameters and markup out of the localization strings.

brylie commented on src/lang/en.json:
```diff
@@ -0,0 +1,391 @@
+          "label": "Requirements Object #$index",
+          "newItemText": "New Security Requirement",
+          "item": {
+            "label": "Security Requirement #$index",
```
My intuition here is telling me that we might only use i18next [interpolation](https://www.i18next.com/interpolation.html) and [pluralization](https://www.i18next.com/plurals.html), leaving most/all other parameters and markup out of the localization strings.

tulir commented:
@brylie I replaced all the form references with interpolations

brylie commented on src/schemas/paths.js:
```diff
@@ -109,48 +112,76 @@ export const paths = {
             'label': 'GET',
             'hideValueIfEmpty': true,
             'isCollapsible': false,
+            'overrideI18nPath': 'form.paths.item.operation',
```
Could we just call this `i18nPath`? It would be a bit shorter, and easier to read. Is `i18nPath` already used?

brylie commented on src/schemas/schemes.js:
```diff
@@ -1,17 +1,11 @@
-    'key': 'wss',
-    'label': 'Secure WebSocket'
-  }]
+  'overrideI18nKeys': {
```
How about just `i18nKeys`?

brylie commented:
In general, we can merge this as-is.

However, I am wondering if we can reduce the verbosity in our i18n keys:

- `overrideI18nPath`
- `overrideI18nKeys`
- `i18nKey`

A couple of minor improvements, to improve consistency, would be:

- remove the word `override` from the keys
- try to consolidate all of those into a single, common, value
  - e.g. `i18n`

tulir commented:
I moved all the generic I18n fields into the `i18n` object and renamed the `i18n` function to `localize`. However, there are still a few problems with unlocalized fields that I need to check out.

tulir commented:
Now everything should work, testing in progress...

brylie commented:
Ah, I merged, since the PR was still marked as ready for review.
